src/reducers/entityReducers/notes.js

```diff
@@ -1,49 +1,128 @@
+  const newArticle = state.getIn(['article', articleIndex]).update('notes', notesAction);
 
-      return state.setIn(['article', articleIndex], newArticle);
-    }
-    case DELETE_NOTE_SUCCESS: {
-      const articleIndex = state.get('article').findIndex((el) => el.get('id') === action.meta.bookmarkId);
+  return state.setIn(['article', articleIndex], newArticle);
+};
 
-      if (articleIndex === MINUS_ONE) {
-        return state;
-      }
+export default (state, action) => {
+  switch (action.type) {
 
-      const newArticle = state.getIn(['article', articleIndex]).update('notes', (old) =>
-        old.filterNot((note) => note.get('id') === action.meta.noteId));
 
-      return state.setIn(['article', articleIndex], newArticle);
+    case CREATE_NOTE_START: {
+      return updateArticle(
+        (old) => old.push(Immutable.fromJS({
+          id: action.meta.note.tmpId,
+          text: action.meta.note.text,
+        })),
+        action.meta.bookmarkId,
+        state,
+      );
+    }
+    case CREATE_NOTE_SUCCESS: {
+      return updateArticle(
+        (old) =>
+          old.map((note) =>
+            (note.get('id') === action.meta.note.tmpId
+              ? Immutable.fromJS({
+                id: action.payload.id,
+                text: action.meta.note.text,
+              })
+              : note)),
+        action.meta.bookmarkId,
+        state,
+      );
+    }
+    case CREATE_NOTE_FAILURE: {
+      return updateArticle(
+        (old) =>
+          old.filterNot((note) =>
+            note.get('id') === action.meta.note.tmpId),
+        action.meta.bookmarkId,
+        state,
+      );
     }
-    case UPDATE_NOTE_SUCCESS: {
-      const articleIndex = state.get('article').findIndex((el) => el.get('id') === action.meta.bookmarkId);
 
-      if (articleIndex === MINUS_ONE) {
-        return state;
-      }
 
-      const newArticle = state.getIn(['article', articleIndex]).update('notes', (old) =>
-        old.map((note) => (note.get('id') === action.meta.noteId ? Immutable.fromJS(action.payload) : note)));
+    case DELETE_NOTE_START: {
+      return updateArticle(
+        (old) =>
+          old.filterNot((note) =>
+            note.get('id') === action.meta.note.id),
+        action.meta.bookmarkId,
+        state
+      );
+    }
+    case DELETE_NOTE_FAILURE: {
+      return updateArticle(
+        (old) => old.splice(
+          action.meta.note.index,
+          ZERO,
+          Immutable.fromJS({
+            id: action.meta.note.id,
+            text: action.meta.note.text,
+          })
+        ),
+        action.meta.bookmarkId,
+        state
+      );
+    }
 
-      return state.setIn(['article', articleIndex], newArticle);
+
+    case UPDATE_NOTE_START: {
+      return updateArticle(
+        (old) =>
+          old.map((note) =>
+            (note.get('id') === action.meta.note.id
+              ? Immutable.fromJS({
+                id: action.meta.note.id,
+                text: action.meta.note.newText,
```

Comment:
Use `action.payload` for newText? Basically, to make it more generic, merge `action.meta.note` with `action.payload`.

Reply:
What?

Reply:
Imagine the case of adding new field to notes.

Reply:
Could you elaborate more? :smile: